src/Common/src/TypeSystem/Common/TypeSystemHelpers.cs

```diff
@@ -36,18 +36,24 @@ static public PointerType MakePointerType(this TypeDesc type)
 
         static public DefType GetClosestDefType(this TypeDesc type)
         {
-            if (type is DefType)
-                return (DefType)type;
-            else
-                return type.BaseType;
+            // The implementation for this should mirror GetClosestMetadataType
+            throw new NotImplementedException();
         }
 
         static public MetadataType GetClosestMetadataType(this TypeDesc type)
         {
-            if (type is MetadataType)
-                return (MetadataType)type;
-            else
-                return type.BaseType.GetClosestMetadataType();
+            if (type.IsSzArray)
+            {
+                MetadataType arrayType = type.Context.SystemModule.GetType("System", "Array`1");
```

Comment:
ResolveInterfaceMethodToVirtualMethodOnType would be virtual. ArrayOfTRuntimeInterfacesAlgorithm.cs would provide own implementation - similar to what it does for the list of interfaces. Do you still see a problem with that?